amazon-ebssurrogate plugin: Added mutex to synchronize map access across threads.

deisenzimmer opened the pull request:
We use packer with the amazon-ebssurrogate plugin to create custom partitioned AMI's which satisfy CIS hardening requirements.  We put several partitions on their own EBS volumes.  We were seeing intermittent failures with output:

Build 'amazon-ebssurrogate' errored: unexpected EOF[0m

Once debug was turned on the true error was found:

fatal error: concurrent map writes

The thread and call stack dump revealed multiple threads attempting to write to a map.  It appears that parallelism is used for the main code path which does EBS snapshot creation.  Since our partition scheme has 8 volumes this meant 8 threads were possibly trying to write to the map simultaneously and luck would determine whether the error happened.

This fix utilizes a sync.Mutex to ensure that only one thread is writing to the map at a time.  This is my first time with go.  I hope this solution is appropriate.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=8432)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=8432)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=8432).</sub>

deisenzimmer commented:
Thank you @azr for reviewing.  I wasn't thinking that the read lock would be needed because I think all read code paths are single threaded.  I could be wrong and I agree it would be good practice.  Would you mind commenting on the new commit?  I changed it to RWMutex and added the read lock.

azr commented:
Hey @deisenzimmer read-write locks exist for performance reasons. Like if a lot of goroutines are reading at the same time there is no need to lock; but then if we want to change a value we need to make sure everyone is done reading. In this case since we really don't need to be performant; I don't think the complexity of using a RWMutex is more valuable than the simplicity of a basic mutex.

> I wasn't thinking that the read lock would be needed because I think all read code paths are single threaded. 

That's a good point; but because you had the race condition which I could not reproduce we probably will have to be safe.

Could you please try to trigger the race condition again but using packer with the race detector enabled ? You can do so using for example `go run -race . build buildfile.json` from the packer directory.

Do you happen to have a simplified buildfile so that we can test this too ?

azr commented on builder/amazon/ebssurrogate/step_snapshot_volumes.go:
```diff
@@ -104,11 +108,13 @@ func (s *StepSnapshotVolumes) Cleanup(state multistep.StateBag) {
 				ui.Error(fmt.Sprintf("Error: %s", err))
 			}
 		}
+		s.snapshotMutex.Unlock()
```
Again I would recommend not using rwlock mutexes but if you think this is better this needs to be changed to 
```suggestion
		s.snapshotMutex.Runlock()
```

azr commented on builder/amazon/ebssurrogate/step_snapshot_volumes.go:
```diff
@@ -104,11 +108,13 @@ func (s *StepSnapshotVolumes) Cleanup(state multistep.StateBag) {
 		ec2conn := state.Get("ec2").(*ec2.EC2)
 		ui := state.Get("ui").(packer.Ui)
 		ui.Say("Removing snapshots since we cancelled or halted...")
+		s.snapshotMutex.RLock() // read lock
```
```suggestion
		s.snapshotMutex.RLock()
```

azr commented on builder/amazon/ebssurrogate/step_snapshot_volumes.go:
```diff
@@ -61,6 +64,7 @@ func (s *StepSnapshotVolumes) Run(ctx context.Context, state multistep.StateBag)
 	ui := state.Get("ui").(packer.Ui)
 
 	s.snapshotIds = map[string]string{}
+	s.snapshotMutex = sync.RWMutex{}
```
A mutex does not need to be initialised.

```suggestion
	s.snapshotMutex = sync.RWMutex{}
```
